src/validation/rules/SingleFieldSubscriptionsRule.js

```diff
@@ -1,30 +1,95 @@
+import type { ObjMap } from '../../jsutils/ObjMap';
 import { GraphQLError } from '../../error/GraphQLError';
 
 import type { ASTVisitor } from '../../language/visitor';
-import type { OperationDefinitionNode } from '../../language/ast';
+import type {
+  OperationDefinitionNode,
+  FragmentDefinitionNode,
+} from '../../language/ast';
+import { Kind } from '../../language/kinds';
 
-import type { ASTValidationContext } from '../ValidationContext';
+import type { ValidationContext } from '../ValidationContext';
+import type { ExecutionContext } from '../../execution/execute';
+import {
+  collectFields,
+  defaultFieldResolver,
+  defaultTypeResolver,
+} from '../../execution/execute';
 
 /**
- * Subscriptions must only include one field.
+ * Subscriptions must only include a non-introspection field.
  *
- * A GraphQL subscription is valid only if it contains a single root field.
+ * A GraphQL subscription is valid only if it contains a single root field and
+ * that root field is not an introspection field.
  */
 export function SingleFieldSubscriptionsRule(
-  context: ASTValidationContext,
+  context: ValidationContext,
 ): ASTVisitor {
   return {
     OperationDefinition(node: OperationDefinitionNode) {
       if (node.operation === 'subscription') {
-        if (node.selectionSet.selections.length !== 1) {
-          context.reportError(
-            new GraphQLError(
-              node.name
-                ? `Subscription "${node.name.value}" must select only one top level field.`
-                : 'Anonymous Subscription must select only one top level field.',
-              node.selectionSet.selections.slice(1),
-            ),
+        const schema = context.getSchema();
+        const subscriptionType = schema.getSubscriptionType();
+        if (subscriptionType) {
+          const operationName = node.name ? node.name.value : null;
+          const variableValues: {
+            [variable: string]: mixed,
+            ...
+          } = {};
+          const document = context.getDocument();
+          const fragments: ObjMap<FragmentDefinitionNode> = Object.create(null);
+          for (const definition of document.definitions) {
+            if (definition.kind === Kind.FRAGMENT_DEFINITION) {
+              fragments[definition.name.value] = definition;
+            }
+          }
+          const fakeExecutionContext: ExecutionContext = {
+            schema,
+            fragments,
+            rootValue: undefined,
+            contextValue: undefined,
+            operation: node,
+            variableValues,
+            fieldResolver: defaultFieldResolver,
+            typeResolver: defaultTypeResolver,
+            errors: [],
+          };
+          const fields = collectFields(
+            fakeExecutionContext,
+            subscriptionType,
+            node.selectionSet,
+            {},
+            {},
```

Comment:
With objects as maps they would count `fragment hasOwnProperty` as visited and exhibit other weird behavior.
In other places we used `Object.create(null)` but that is confusing as your PR shows, so I changed `collectFields` to accept `Map` and `Set`.
@benjie Can you rebase your PR?

Reply:
Done :+1: